perf(ngcc): implement entry-point manifest

petebacondarwin opened the pull request:
The `DirectoryWalkerEntryPointFinder` has to traverse the
entire node_modules library everytime it executes in order to
identify the entry-points that need to be processed. This is
very time consuming (several seconds for big projects on
Windows).

This commit changes the `DirectoryWalkerEntryPointFinder` to
use the `EntryPointManifest` to store the paths to entry-points
that were found when doing this initial node_modules traversal
in a file to be reused for subsequent calls.

This dramatically speeds up ngcc processing when it has been run once already.

FW-1962

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * Manages reading a write a manifest file that contains a list of all the entry-points that were
+ * found below a give basePath.
```
```suggestion
 * Manages reading and writing a manifest file that contains a list of all the entry-points that were
 * found below a given basePath.
```

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+ * Manages reading a write a manifest file that contains a list of all the entry-points that were
+ * found below a give basePath.
+ *
+ * This is as super-set of the entry-points that are actually processed for a given run of ngcc,
```
```suggestion
 * This is a super-set of the entry-points that are actually processed for a given run of ngcc,
```

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+        }
+      }
+      const duration = Math.round((Date.now() - startTime) / 100) / 10;
+      this.logger.debug(`Reading entry-points directly took ${duration}s.`);
```
nit: Mentioning _manifest_ here is more clear than _directly_

```suggestion
      this.logger.debug(`Reading entry-points from manifest took ${duration}s.`);
```

petebacondarwin replied:
Strictly the entry-points were not read from the manifest, only their paths are read from the manifest. We still read the entry-point info from the files on disk.

petebacondarwin replied:
```suggestion
      this.logger.debug(`Reading entry-points using the manifest entries took ${duration}s.`);
```

JoostK replied:
Oh, yeah that is true. I was thinking that if I were to read that debug message, I wouldn't know what _directly_ would actually mean.

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+      this.logger.debug(`Reading entry-points directly took ${duration}s.`);
+      return entryPoints;
+    } catch (e) {
+      this.logger.warn(`Unable to read the entry-point manifest for ${basePath}`, e.toString());
```
I would include a colon before the error message:

```suggestion
      this.logger.warn(`Unable to read the entry-point manifest for ${basePath}: ${e.toString()}`);
```

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+      lockFileHash: lockFileHash,
+      entryPointPaths: entryPoints.map(entryPoint => [entryPoint.package, entryPoint.path]),
+    };
+    this.fs.writeFile(this.getEntryPointManifestPath(basePath), JSON.stringify(manifest));
```
Consider writing this as readable JSON:
```suggestion
    this.fs.writeFile(this.getEntryPointManifestPath(basePath), JSON.stringify(manifest, null, 2));
```

petebacondarwin replied:
I avoided this because:

a) It is not supposed to be read by humans. And if they want to read it, it is trivial to format it in an editor
b) These files can get quite large. Adding extra whitespace slows down the stringification/parsing for now programmatic benefit. Given that this change is supposed to be for performance reasons that seemed a strange thing to do.

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+    return this.fs.resolve(basePath, '__ngcc_entry_points__.json');
+  }
+
+  private computeLockFileHash(basePath: AbsoluteFsPath): string|null {
```
Should this be cached? Otherwise it would have to be computed twice if there's no manifest file.

petebacondarwin replied:
That is an interesting idea...

* In the case we are trying to optimise (when the manifest does exist) there would be no benefit.
* In the other case it would speed things up. By how much I don't know:
  * The FileSystem will have cached the contents of the file so there would be no actual file access
  * I don't know how time consuming the hash computation is but I expect it is done in C++ code so is probably pretty fast.

I am a bit worried about memoizing this sort of thing... What happens if there happens to be two separate `basePath`s that are `node_modules` in the processing - we would need to store a lookup of lockfile hash per basePath. The complexity seems over-kill.

I will run a few performance checks to see what the cost of this is...

petebacondarwin replied:
On my laptop, running in `ngcc -l debug` in the `aio` folder I get:

```
Debug: Analyzing entry-points...
computeLockFileHash: 2.956ms
Debug: No manifest found for /Users/pete/dev/angular/tmp/ngcc-analysis-manifest/aio/node_modules so walking the directories for entry-points.
...
Debug: Walking directories took 1.4s.
computeLockFileHash: 0.919ms
```

And then when running a second time:

```
Debug: Analyzing entry-points...
computeLockFileHash: 2.769ms
Debug: Entry-point manifest found for /Users/pete/dev/angular/tmp/ngcc-analysis-manifest/aio/node_modules so loading entry-point information directly.
Debug: Reading entry-points using the manifest entries took 0.1s.
```

So it seems that we would, at the most, save 2ms by caching the hash, which I think we can agree is not worth the additional complexity.

JoostK replied:
That's not worth the complexity, indeed! Thanks for looking into it 👍

JoostK commented on packages/compiler-cli/ngcc/test/packages/entry_point_manifest_spec.ts:
```diff
@@ -0,0 +1,238 @@
+      it('should return null if the lockfile hash does not match', () => {
+        fs.ensureDir(_Abs('/project/node_modules'));
+        fs.writeFile(_Abs('/project/yarn.lock'), 'LOCK FILE CONTENTS');
+        manifestFile.lockFileHash = 'bad-version';
```
```suggestion
        manifestFile.lockFileHash = 'bad-hash';
```

JoostK commented on packages/compiler-cli/ngcc/test/packages/entry_point_manifest_spec.ts:
```diff
@@ -0,0 +1,238 @@
+        expect(entryPoints).toBe(null);
+      });
+
+      it('should return an array of entry-points if all the checks pass', () => {
```
Could you duplicate this test for `package-lock.json`?

JoostK commented on packages/compiler-cli/ngcc/test/packages/entry_point_manifest_spec.ts:
```diff
@@ -0,0 +1,238 @@
+        expect(entryPoints).toBe(null);
+      });
+
+      it('should return null if the lockfile hash does not match', () => {
```
Could you duplicate this test for `package-lock.json`?

petebacondarwin commented:
Marking for merge but waiting to see if @gkalpak or @alxhub want to comment.

petebacondarwin commented:
Add an `--invalidate-manifest` flag

petebacondarwin committed:
refactor(ngcc): expose a hash of the project configuration

This will be used in the entry-point manifest since a change to
configuration might change the entry-points that are found.

petebacondarwin committed:
refactor(ngcc): show timings in 1/10ths of a second

petebacondarwin committed:
refactor(ngcc): add entry-point manifest functionality

The new `EntryPointManifest` class can read and write a
manifest file that contains all the paths to the entry-points
that have been found in a node_modules folder.
This can be used to speed up finding entry-points in
subsequent runs.

The manifest file stores the ngcc version and hashes of
the package lock-file and project config, since if these
change the manifest will need to be recomputed.

petebacondarwin committed:
feat(ngcc): support invalidating the entry-point manifest

In some scenarios it is useful for the developer to indicate
to ngcc that it should not use the entry-point manifest
file, and instead write a new one.

In the ngcc command line tool, this option is set by specfying

```
--invalidate-entry-point-manifest
```

petebacondarwin committed:
perf(ngcc): use the `EntryPointManifest` in `DirectoryWalkerEntryPointFinder`

The `DirectoryWalkerEntryPointFinder` has to traverse the
entire node_modules library everytime it executes in order to
identify the entry-points that need to be processed. This is
very time consuming (several seconds for big projects on
Windows).

This commit changes the `DirectoryWalkerEntryPointFinder` to
use the `EntryPointManifest` to store the paths to entry-points
that were found when doing this initial node_modules traversal
in a file to be reused for subsequent calls.

This dramatically speeds up ngcc processing when it has been run once
already.

petebacondarwin commented:
Added the new `--invalidate-entry-point-manifest` flag. PTAL @JoostK et al

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -128,6 +128,18 @@ export class EntryPointManifest {
+ * A specialized implementation of the `EntryPointManifest` that always returns `null` rather than a
+ * set of entry-points from a manifest file.
+ *
+ * This is used if you want to ignore a manifest file that is written to disc.
```
```suggestion
 * This is used if you want to ignore a manifest file that is written to disk.
```

Or are you holding on to portable media 💿😄

JoostK commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -128,6 +128,18 @@ export class EntryPointManifest {
+ * Note that it will still write a new manifest file with the entry-points that were found by
+ * walking the directories.
+ */
+export class NullEntryPointManifest extends EntryPointManifest {
```
May I propose `IgnoredEntryPointManifest`, as it's not entirely a null implementation.

```suggestion
export class IgnoredEntryPointManifest extends EntryPointManifest {
```

petebacondarwin replied:
Can I counter propose `InvalidatingEntryPointManifest`, since that is what it is doing?

JoostK replied:
Sure, go ahead 👍

JoostK commented on packages/compiler-cli/ngcc/main-ngcc.ts:
```diff
@@ -76,6 +76,13 @@ if (require.main === module) {
+          .option('invalidate-entry-point-manifest', {
+            describe:
+                'If this is set then ngcc will not read an entry-point manifest file from disk.\n' +
+                'Instead it will walking the directory tree as normal looking for entry-points, and then write a new manifest file.',
```
```suggestion
                'Instead it will walk the directory tree as normal looking for entry-points, and then write a new manifest file.',
```

JoostK commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -982,6 +982,39 @@ runInEachFileSystem(() => {
+        commonTesting =
+            JSON.parse(fs.readFile(_('/node_modules/@angular/common/testing/package.json')));
+        expect(hasBeenProcessed(commonTesting, 'esm5')).toBe(true);
+        expect(hasBeenProcessed(commonTesting, 'esm2015')).toBe(true);
```
Consider adding an assertion that the manifest file has been updated with the newly scanned entry-points.

petebacondarwin replied:
Done

alan-agius4 commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,151 @@
+ * This is a super-set of the entry-points that are actually processed for a given run of ngcc,
+ * since some may already be processed, or excluded if they do not have the required format.
+ */
+export class EntryPointManifest {
```
Nit: for consistence, should you always use `EntryPoints` or `EntryPoint`

- readEntryPointsUsingManifest
- writeEntryPointManifest
- getEntryPointManifestPath
- invalidateEntryPointManifest

I think case, probably it's probably it's prefer the plural versions, since you have a manfest which contains entry-points data.

petebacondarwin replied:
Actually I believe it is already consistent. The manifest is an "entry-point manifest" (like at school you might have a "student register" not a "students register"). It is referred in this way consistently throughout, unless I have missed something. I feel the name `EntryPointsManifest` is overly verbose.

The only difference is the `readEntryPointsUsingManifest()` method but this also is consistent: it is reading "entry-points" using the "entry-point manifest" as a shortcut.

Let me know if there are other places that are inconsistent with this naming.

alan-agius4 replied:
Thanks for the detailed explanation.

petebacondarwin committed:
fixup! feat(ngcc): support invalidating the entry-point manifest

petebacondarwin committed:
fixup! feat(ngcc): support invalidating the entry-point manifest

petebacondarwin committed:
fixup! feat(ngcc): support invalidating the entry-point manifest

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+   *
+   * * the file does not exist
+   * * the ngcc version has changed
+   * * the package lock-file (i.e. yarn.lock or package-lock.json) has changed
```
Wouldn't `project` (instead of `package`) be more consistent with how we use it when referring to ngcc configurations?
(Here and in other method docs below.)

petebacondarwin replied:
By "package lock-file" I wanted to indicate that we are talking about file that locks down the package versions - rather than our other lock-file which is used to lock down the current running instance of ngcc. IMO project would be even more ambiguous.

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+      const computedLockFileHash = this.computeLockFileHash(basePath);
+      if (computedLockFileHash === null) {
+        return null;
+      }
```
Nit: I would move this below the `mnifestPath` check, because this check it more expensive and less likely to fail (if the next one succeeds).

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+   * * the ngcc version has changed
+   * * the package lock-file (i.e. yarn.lock or package-lock.json) has changed
+   * * the project configuration has changed
+   * * one or more entry-points in the manifest are not valid
```
AFAICT, this case results in an error being thrown (not the manifest being considered "out-of-date").

petebacondarwin replied:
Note the big try-catch block around the whole function. It never throws but returns `null` if there was an error.

gkalpak replied:
Ooops, missed that :grin:

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point_manifest.ts:
```diff
@@ -0,0 +1,139 @@
+      this.logger.debug(`Reading entry-points using the manifest entries took ${duration}s.`);
+      return entryPoints;
+    } catch (e) {
+      this.logger.warn(`Unable to read the entry-point manifest for ${basePath}:`, e.toString());
```
Why not just pass `e`? `e.toString()` loses the stack trace info, which might be helpful in debugging.

petebacondarwin replied:
`Logger.warn()` expects strings

petebacondarwin replied:
We could do `e.stack` possibly, falling back on `e.toString()` if there is none.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -117,6 +117,17 @@ export interface SyncNgccOptions {
+
+  /**
+   * Whether to read an entry-point manifest file from disk, rather than walking the directory tree
+   * looking for entry-points.
```
This reads backwards to me 😇
How about inverting it:

```suggestion
   * Whether to walk the directory tree looking for entry-points rather than trying to read an
   * entry-point manifest file from disk.
```

petebacondarwin replied:
Yes, looking back this doesn't read well.

petebacondarwin replied:
Going for:

```
   * Whether to invalidate any entry-point manifest file that is on disk. Instead, walk the
   * directory tree looking for entry-points, and then write a new entry-point manifest, if
   * possible.
```

gkalpak reviewed (approved):
I am a little late to the party, but since I had started the review yesterday 🤷‍♂
Nothing blocking, so feel free to just add the `merge` label back without addressing my comments 😇

petebacondarwin commented on packages/compiler-cli/ngcc/src/execution/cluster/master.ts:
```diff
@@ -70,7 +70,7 @@ export class ClusterMaster {
     // First, check whether all tasks have been completed.
     if (this.taskQueue.allTasksCompleted) {
-      const duration = Math.round((Date.now() - this.processingStartTime) / 1000);
+      const duration = Math.round((Date.now() - this.processingStartTime) / 100) / 10;
```
Feel free to do a refactor PR after this lands to clean this up.

petebacondarwin committed:
fixup! refactor(ngcc): add entry-point manifest functionality

petebacondarwin commented:
Thanks @gkalpak .
